src/stream/src/executor/hash_agg.rs

```diff
@@ -432,57 +472,85 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
             .inc_by(*chunk_total_lookup_count);
         *chunk_total_lookup_count = 0;
 
-        let dirty_cnt = group_change_set.len();
+        let dirty_cnt = held_keys.len();
         if dirty_cnt > 0 {
             // Produce the stream chunk
             let group_key_data_types = &schema.data_types()[..group_key_indices.len()];
-            let mut group_chunks = IterChunks::chunks(group_change_set.drain(), *chunk_size);
-            while let Some(batch) = group_chunks.next() {
-                let keys_in_batch = batch.into_iter().collect_vec();
+            let mut to_flush_keys = Vec::with_capacity(held_keys.len());
+            let mut to_emit_chunks = IterChunks::chunks(
+                held_keys
+                    .drain_filter(|k, dirty| {
+                        if *dirty {
+                            to_flush_keys.push(k.clone());
+                        }
+                        *dirty = false;
 
-                // Flush agg states.
-                for key in &keys_in_batch {
-                    let agg_group = agg_group_cache
-                        .get_mut(key)
-                        .expect("changed group must have corresponding AggGroup")
-                        .as_mut();
-                    agg_group.flush_state_if_needed(storages).await?;
-                }
+                        let key = K::deserialize(k, group_key_data_types).unwrap();
+                        E::should_emit(key, &buffered_watermarks)
+                    })
+                    .map(|(k, _)| k),
+                *chunk_size,
+            );
+            while let Some(batch) = to_emit_chunks.next() {
+                let keys_in_batch = batch.into_iter().collect_vec();
 
                 // Create array builders.
                 // As the datatype is retrieved from schema, it contains both group key and
                 // aggregation state outputs.
                 let mut builders = schema.create_array_builders(chunk_size * 2);
                 let mut new_ops = Vec::with_capacity(chunk_size * 2);
 
+                let agg_groups = {
+                    let mut agg_group_cache_ptr: NonNull<_> = agg_group_cache.into();
+                    keys_in_batch
+                        .iter()
+                        .map(|key| {
+                            // SAFETY: `keys_in_batch` is a subset of `held_keys`, which is a
+                            // `HashMap`, so we can ensure that every `&mut agg_group_cache` is
+                            // unique.
+                            unsafe {
```

Comment:
Similar to https://doc.rust-lang.org/std/primitive.slice.html#method.split_at_mut, this is definitely a safe behavior.

Reply:
Is this magic just to avoid `agg_group_cache.pop` and `agg_group_cache.put`? It looks OK to me but is it worth doing so? I mean in terms of maintainability and the fact that it does add an `unsafe` block.

Reply:
Makes sense.

Reply:
Makes sense.